io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -527,6 +533,18 @@ schedule thread@Thread{
     -- ExploreRaces is ignored by this simulator
     ExploreRaces k -> schedule thread{ threadControl = ThreadControl k ctl } simstate
 
+    Fix f k -> do
+      m <- newSTRef (throw NonTermination)
+      x <- unsafeInterleaveST $ readSTRef m
+      let -- recursion in 'IOSim'
+          io = do
+            x' <- f x
+            -- note: the bang is crucial!
+            !_ <- liftST $ lazyToStrictST $ writeSTRef m x'
+            return x'
+          thread' = thread { threadControl = ThreadControl (unIOSim io k) ctl }
+      schedule thread' simstate
```

Comment:
Since this is low level voodoo, I'd be super-explicit and not use wrappers like `liftST` and go direct to the constructors, e.g.
```suggestion
    Fix f k -> do
      r <- newSTRef (throw NonTermination)
      x <- unsafeInterleaveST (readSTRef r)
      let k'      = unIOSim (f x) $ \x' ->
                      LiftST (lazyToStrictST (writeSTRef r x')) (\() -> k x')
          thread' = thread { threadControl = ThreadControl k' ctl }
      schedule thread' simstate
```

Reply:
:+1: